File: epms/src/services/grReport.ts
import { api, downloadFile } from '@/lib/api'

/** Today as YYYY-MM-DD in the reader's own zone (not UTC). */
function localDay(): string {
  const d = new Date()
  return `${d.getFullYear()}-${String(d.getMonth() + 1).padStart(2, '0')}-${String(d.getDate()).padStart(2, '0')}`
}

/**
 * Warehouse receiving report — one row per physically received line item.
 *
 * Dates arrive as plain `YYYY-MM-DD` strings already resolved to the plant's
 * local day by the API. Render them with `formatDateOnly` / as-is; never through
 * `new Date(...)`, which reads a date-only string as UTC midnight and shows the
 * previous day in a UTC-4 browser.
 */
export interface ReceivingReportRow {
  gr_id: string
  gr_number: string
  po_id: string
  po_number: string
  material_id: string | null
  description: string
  supplier: string
  unit: string
  quantity: string
  department: string | null
  requested_by: string | null
  date_ordered: string | null
  arrival_date: string | null
  left_warehouse_date: string | null
  warehouse_receiver: string | null
  person_accepting: string | null
  lead_time_days: number | null
}

export interface ReceivingReportResponse {
  items: ReceivingReportRow[]
  total: number
  truncated: boolean
}

export interface ReceivingReportFilters {
  // Index signature so the object satisfies the query-param type `api.get` and
  // `downloadFile` take; without it every call site needs a cast.
  [key: string]: string | undefined
  date_from?: string
  date_to?: string
  department_id?: string
  vendor_id?: string
  search?: string
}

export const grReportService = {
  receiving: (filters?: ReceivingReportFilters) =>
    api.get<ReceivingReportResponse>('/gr/receiving-report', filters),

  exportReceiving: (filters?: ReceivingReportFilters) =>
    // The filename is passed explicitly: EPMS talks to epms-api cross-origin,
    // and a browser hides Content-Disposition from such a response unless the
    // server exposes it — leaving the shared helper to fall back on
    // "export.csv", an xlsx body under a CSV name that Excel then refuses.
    downloadFile('/gr/receiving-report/export', filters, `receiving-report-${localDay()}.xlsx`),
}

File: epms-api/app/schemas/gr_report.py
"""Pydantic schemas for the warehouse receiving report."""
import uuid
from datetime import date
from decimal import Decimal

from pydantic import BaseModel


class ReceivingReportRow(BaseModel):
    gr_id: uuid.UUID
    gr_number: str
    po_id: uuid.UUID
    po_number: str
    material_id: str | None
    description: str
    supplier: str
    unit: str
    quantity: Decimal
    department: str | None
    requested_by: str | None
    # Plain dates, already resolved to the plant's local day by the query layer.
    # The front-end renders them as-is: parsing a date-only string through
    # `new Date()` in a UTC-4 browser lands on the previous day.
    date_ordered: date | None
    arrival_date: date | None
    left_warehouse_date: date | None
    warehouse_receiver: str | None
    person_accepting: str | None
    lead_time_days: int | None

    model_config = {"from_attributes": True}


class ReceivingReportResponse(BaseModel):
    items: list[ReceivingReportRow]
    total: int
    # True when the row cap cut the result short — the UI says so rather than
    # letting a half-answer look complete.
    truncated: bool
